framework/src/utils/BootstrapStatistics.C

```diff
@@ -77,20 +80,77 @@ BootstrapCalculator::shuffle(const std::vector<Real> & data,
                              MooseRandom & generator,
                              const bool is_distributed) const
 {
-  const std::size_t n = data.size();
-  std::vector<Real> replicate(n);
+  // Size of the local input data
+  const std::size_t n_local = data.size();
+
+  // The data to be returned
+  std::vector<Real> replicate;
+
+  // REPLICATED data
   if (!is_distributed)
   {
-    for (std::size_t j = 0; j < n; ++j)
+    replicate.resize(n_local);
+    for (std::size_t j = 0; j < n_local; ++j)
     {
-      auto index = generator.randl(0, 0, n);
+      auto index = generator.randl(0, 0, n_local);
       replicate[j] = data[index];
     }
   }
 
+  // DISTRIBUTED data
   else
   {
-    // from future import parallel magic
+    // Reserve space for storing the local
+    replicate.reserve(n_local);
+
+    // Compute the global size of the vector
+    std::size_t n_global = n_local;
+    _communicator.sum(n_global);
+
+    // Compute the vector data offsets, the scope cleans up the "n_local" vector
+    std::vector<std::size_t> offsets(n_processors());
+    {
+      std::vector<std::size_t> local_sizes;
+      _communicator.allgather(n_local, local_sizes);
+      for (std::size_t i = 1; i < local_sizes.size(); ++i)
+        offsets[i] = offsets[i - 1] + local_sizes[i];
+    }
+
+    // Advance the random number generator to the current offset
+    for (std::size_t i = 0; i < offsets[processor_id()]; ++i)
+      generator.randl(0, 0, n_global);
+
+    // Compute the needs for this processor
+    std::unordered_map<processor_id_type, std::vector<std::size_t>> needs;
+    for (std::size_t i = 0; i < n_local; ++i)
+    {
+      auto index = generator.randl(0, 0, n_global);
+      // iterator to first element in range that is GREATER THAN index
+      auto rank = std::upper_bound(offsets.begin(), offsets.end(), index);
+      auto rk = std::distance(offsets.begin(), rank) - 1;
+      needs[rk].push_back(index - offsets[rk]);
+    }
+
+    // Advance the random number generator to the end of the global vector
+    for (std::size_t i = offsets[processor_id() + 1]; i < n_global; ++i)
```

Comment:
```suggestion
    for (std::size_t i = offsets[processor_id()] + n_local; i < n_global; ++i)
```

This'll fix your crash